test/HaveLetLinter.lean

```diff
@@ -0,0 +1,144 @@
+import Mathlib.Tactic.Linter.HaveLetLinter
+import Mathlib.Tactic.Tauto
+
+/-- a tactic that simply logs an empty message.  Useful for testing the chattiness of the
+`haveLet` linter. -/
+elab "noise" : tactic => do Lean.logInfo ""
+
+set_option linter.haveLet 2 in
+#guard_msgs in
+-- check that `tauto`, `replace`, `classical` are ignored
+example : True := by
+  classical
+  let zero := 0
+  replace _zero := zero
+  let eq := (rfl : 0 = 0)
+  replace _eq := eq
+  tauto
+
+/--
+info:
+
+---
+warning: '_zero : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'.
+[linter.haveLet]
+-/
+#guard_msgs in
+example : True := by
+  noise
+  have ⟨_zero, _⟩ : Fin 1 := ⟨0, Nat.zero_lt_one⟩
+  exact .intro
+
+-- I don't know how to silence this test.
+--set_option linter.haveLet 2 in
+--/--
+--warning: '_zero : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'.
+--[linter.haveLet]
+---/
+--#guard_msgs in
+--example : True := by
+--  have ⟨_zero, _⟩ : Fin 1 := ⟨0, Nat.zero_lt_one⟩
+--  exact .intro
+
+#guard_msgs in
+example : True := by
+  have ⟨_zero, _⟩ : Fin 1 := ⟨0, Nat.zero_lt_one⟩
+  exact .intro
+
+/--
+info:
+
+---
+warning: '_zero : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'.
+[linter.haveLet]
+-/
+#guard_msgs in
+example : True := by
+  have ⟨_zero, _⟩ : Fin 1 := ⟨0, Nat.zero_lt_one⟩
+  noise
+  exact .intro
+
+/--
+info:
+
+---
+warning:
+'_zero : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'. [linter.haveLet]
+-/
+-- when `have` introduces several hypotheses, the linter flags the non-`Prop` ones.
+#guard_msgs in
+example : True := by
+  have ⟨_zero, _⟩ : Fin 1 := ⟨0, Nat.zero_lt_one⟩
+  noise
+  exact .intro
+
+/--
+info:
+
+---
+warning: '_a : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'.
+[linter.haveLet]
+---
+warning: '_b : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'.
+[linter.haveLet]
+---
+warning: '_oh : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'.
+[linter.haveLet]
+---
+warning: '_b : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'.
+[linter.haveLet]
+-/
+#guard_msgs in
+example : True := by
+  noise
+  have _a := 0
+  have _b : Nat := 0
+  have _b : 0 = 0 := rfl
+  have _oh : Nat := 0
+  have _b : Nat := 2
+  tauto
```

Comment:
Is there a way to guard an unfinished proof, like this one without `tauto`? If I understand your motivation for level `1` correctly, it's also to trigger on unfinished proofs, isn't it?

I expected that if I delete the `#guard_msgs` here and then delete `tauto` that the linter warnings would pop up. Did I understand this wrongly?

Reply:
ok. I somehow expect that an "unsolved goals" error behaves very similar to a "declaration uses sorry" warning in as many aspects as possible but I see that's not as easy with how messages are reported.

Looks good to me as is.